docs/DesignDocs/AttestationAPIsWithFlexibleClaims.md

```diff
@@ -0,0 +1,571 @@
+            claims_length);
+    }
+
+    OE_CHECK(result);
+    OE_TRACE_VERBOSE("quote validation succeeded");
+
+    // verify report data: hash(public key)
+    // extract public key from the cert
+    oe_memset_s(public_key_buffer, KEY_BUFF_SIZE, 0, KEY_BUFF_SIZE);
+    result = oe_cert_write_public_key_pem(
+        &cert, public_key_buffer, &public_key_buffer_size);
+    OE_CHECK(result);
+    OE_TRACE_VERBOSE(
+        "oe_cert_write_public_key_pem pub_key_buf_size=%d",
+        public_key_buffer_size);
+
+    result = _verify_public_key_claim(
+        *claims, *claims_length, public_key_buffer, public_key_buffer_size);
+    OE_CHECK(result);
+    OE_TRACE_VERBOSE("user data: hash(public key) validation passed", NULL);
+
++   oe_claim_t* custom_claims_buffer;
++   if (_find_claim(
++           *claims,
++           *claims_length,
++           OE_CLAIM_CUSTOM_CLAIMS_BUFFER,
++           &custom_claims_buffer) == OE_OK)
++   {
++       if (custom_claims_buffer->value_size > public_key_buffer_size)
++       {
++           size_t runtime_custom_claims_size =
++               custom_claims_buffer->value_size - public_key_buffer_size;
++           uint8_t* runtime_custom_claims =
++               (uint8_t*)((uint64_t)custom_claims_buffer->value + public_key_buffer_size);
++           _add_claim(
++               claims,
++               claims_length,
++               OE_CLAIM_RUNTIME_CUSTOM_CLAIMS_BUFFER,
++               sizeof(OE_CLAIM_RUNTIME_CUSTOM_CLAIMS_BUFFER),
++               runtime_custom_claims,
++               runtime_custom_claims_size);
++       }
++   }
+
+done:
+    oe_free(public_key_buffer);
+    oe_cert_free(&cert);
+    oe_free(report);
+    return result;
+}
+```
+
+Discussion
+--------
+
+- Separation of custom claims and run-time custom claims.
+
+  The current design introduces a separated `OE_CLAIM_RUNTIME_CUSTOM_CLAIMS_BUFFER`
+  in addition to `OE_CLAIM_CUSTOM_CLAIMS_BUFFER`. Doing so allows for keeping
+  the existing behavior of verifying public and providing more flexible use
+  of run-time custom claims. The downside is confusing the use of custom claims.
+  An alternative is
+  - Rename the `OE_CLAIM_CUSTOM_CLAIMS_BUFFER` as `OE_CLAIM_RUNTIME_CUSTOM_CLAIMS_BUFFER`
+  - Introduce `OE_CLAIM_RUNTIME_CUSTOM_CLAIMS_PUBLIC_KEY` (or simply `OE_CLAIM_CUSTOM_CLAIMS_PUBLIC_KEY`)
+
+  This topic is open for discussion.
+
+- Verification of init-time custom claims.
+
+  The current design does not implement the logic of verifying init-time custom claims.
+  This decision is letting the developers implement such logic. 
+  Alternatively, the implementation can enforce the check of hash(`inittime_custom_claims`) against
+  `CONFIGID` by default. This topic is also open for discussion.
+
+- Lack of integrity protection with init-time custom claims.
+
+  Based on the current design, the run-time custom claims are integrity-protected (i.e.,
+  its hash is part of the SGX report) while init-time custom claims rely on the
```

Comment:
Why is this specific to "SGX reports"?   This design doc should be support multiple evidence formats.